### PR TITLE
sessions and middleware

### LINKED ISSUE #502: Sessions
Sessions as they currently are in Vapor are not currently working as intended.

Two things:

**Sessions currently only last for the browser session**

This is because the session cookie doesn't have an expiration date, so they are deleted by the browser when it closes. Talking with @tannernelson, this should last forever.

**Sessions don't set a path for the cookie**

Because sessions don't set a path for the cookie, a session cookie initiated on `/login/facebook` or some other subpath will only be valid for that path and subpaths. The cookie path should be `/`

### PR DESCRIPTION
**Pull request by tannernelson**
Adds more configurability to middleware and moves sessions to its own module.

Fixes #502 #559 #419 #579

Todo:
- [x] Add more tests
- [x] Discuss `vapor-auth` being separate cookie token
- [x] ~~Move middleware to HTTP server~~ (would break other stuff)
- [x] Support client _and_ server middleware in `middleware.json`

### Sessions

Sessions are set like so:

```swift
drop.get("set") { req in
    try req.session().data["foo"] = "bar"
    try req.session().data["bar"] = "baz"
    return "set"
}
```

and fetched like so:

```swift
drop.get("get") { req in
    return try req.session().data["foo"]?.string ?? "fail"
}
```

They rely on Node under the hood.

There is a `MemorySessions` and `CacheSessions` conformance out of the box. Memory uses an array and cache uses an underlying `CacheProtocol` to store the sessions. Notably, this could use the same `CacheProtocol` as `Auth`, but it _does not_ use the same cookie.

I think leaving this separate allows for better modularity, especially since Turnstile replicates some of the functionality of `SessionsProtocol`.

Maybe these will be separate modules some day that can be optional. Probably best they don't rely on each other.

### Middleware

Middleware is  now a `let` constant on Droplet and is supplied as a `[String: Middleware]` dict to the init or through providers. This allows for the provided middleware (both from the init and from providers) to be selected at will from the `Config/middleware.json` file. This will be great for choosing different middleware for different environments and also making it clear which middleware is on by default.

Should no `Config/middleware.json` file be present, all provided middleware (both from the init and from providers) will be enabled.

The merging of middleware is unique for providers. Every other property on providers will throw an error if provided by more than one provider. Obviously middleware works a bit differently. Instead, an error will only be thrown if different providers provide a middleware with the same name key.

Additionally, since we have the middleware at init time now, we can pass it directly to the `HTTP.Server` when initialized. That let's Engine (well tested now) handle the middleware logic.

### CODE REVIEW

File: Sources/Sessions/CacheSessions.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import Cache
+import Node
+
+public final class CacheSessions: SessionsProtocol {
+    public let cache: CacheProtocol
+    public init(cache: CacheProtocol) {
+        self.cache = cache
+    }
+
+    public func get(for identifier: String) throws -> Node? {
```
Comment (LoganWright) [inline Sources/Sessions/CacheSessions.swift]: I think subscript here (and set) would be pretty nice.

File: Sources/Sessions/Session.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            do {
+                if let data = try sessions.get(for: i) {
+                    return data
+                } else {
+                    let new = Node([:])
+                    try sessions.set(new, for: i)
+                    return new
+                }
+            } catch {
+                print("[Sessions] Error getting data: \(error)")
```
Comment (LoganWright) [inline Sources/Sessions/Session.swift]: Looking at this, and thinking about new stuff like the SwiftyBeaver connection, I think I'd want to know if this was happening. Maybe we should pump this through Log. I think it'd be awkward to pass in Droplet or something to here, so just wanted to comment so we can keep it in the back of our minds going forward. Global log might do the trick to make sure this stuff comes out.

File: Sources/Sessions/SessionsMiddleware.swift [hunk 3]
```diff
@@ -9,32 +11,33 @@ import Cookies
...
         if
-            let identifier = request.cookies["vapor-session"],
-            sessions.contains(identifier: identifier)
+            let identifier = request.cookies[cookieName],
+            try sessions.contains(identifier)
         {
-            request.session = Session(identifier: identifier, sessions: sessions)
-        } else {
-            request.session = Session(sessions: sessions)
+            session.identifier = identifier
```
Comment (LoganWright) [inline Sources/Sessions/SessionsMiddleware.swift]: If sessions doesn't contain the identifier, but the cookie with the identifier exists, should we still set it? This seems right, just verifying.

File: Sources/Sessions/SessionsProtocol.swift [hunk 4]
```diff
@@ -17,30 +19,34 @@ public protocol Sessions: class {
...
+}
 
+extension SessionsProtocol {
     /**
         Destroy the session associated with given identifier
 
         - parameter identifier: id of session
      */
-    func destroy(_ identifier: String)
+    func destroy(_ identifier: String) throws {
```
Comment (LoganWright) [inline Sources/Sessions/SessionsProtocol.swift]: We should stick `destroy` and `contains` up on the protocol definition so they can be overridden. They'll also need to be public to be accessible

File: Sources/Vapor/Droplet/Droplet.swift [hunk 5]
```diff
@@ -298,22 +319,40 @@ public class Droplet {
...
+            for name in enabled {
+                if let m = middleware[name] {
+                    em.append(m)
+                }
+            }
+        } else if let array = config["middleware"]?.array {
+            // add all middleware specified by
+            // config files
+            for item in array {
+                if let name = item.string {
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: `, let m = middleware[name]`

File: Sources/Vapor/Provider/Providable+Merged.swift [hunk 6]
```diff
@@ -13,6 +13,22 @@ extension Providable {
         func attempt<T: Any>(_ existing: T?, _ new: T?) throws -> T? {
```
Comment (LoganWright) [inline Sources/Vapor/Provider/Providable+Merged.swift]: Not part of this PR, but `attempt` doesn't read very clearly. Looking at this, I was surprised to se it casint ag middleware dictionaries internally. I think the signature of this might need to change. It's not entirely clear why `[String: Middleware]` isn't just used here.

**Comment by LoganWright** [comment]
I'm wondering if we can use a different name than `Sessions`, I think it's somewhat confusing that it's not an array, `(Bytes, Byte)`. I also found it hard to read `Session(sessions: sessions)`. I think originally it was called `SessionDriver`, and I think it was easier for me to understand what the role of the object was. Can we switch back to that, or something that's a little more obvious

File: Sources/Sessions/CacheSessions.swift (hunk 1 above)
Comment (tannernelson) [inline Sources/Sessions/CacheSessions.swift]: Needs to throw.

**Comment by tannernelson** [comment]
It would look like:

```swift
Session(sessions: MemorySessions())
```

You should never have something conforming to `SessionsProtocol` that doesn't have a better descriptor like `CacheSessions` or `MemorySessions`. 

In that case `MemorySessionsDriver` seems excessive.

**Comment by LoganWright** [comment]
@tannernelson I think even `Session(sessions` is a little confusing having the added `s` meaning a pretty different thing. Is there a reason `SessionDriver` wouldn't fit? Something like `Session(driver: MemorySessionDriver())`.

File: Sources/Sessions/CacheSessions.swift (hunk 1 above)
Comment (LoganWright) [inline Sources/Sessions/CacheSessions.swift]: I keep thinking subscripts can do that, good call 👍
